File: behaviorial/template-method/python/abstract.py
from abc import ABC, abstractmethod


class ReportGenerator(ABC):
    def generate_report(self):
        self.collect_data()
        self.process_data()
        self.format_data()
        self.print_report()

    @abstractmethod
    def collect_data(self):
        pass

    @abstractmethod
    def process_data(self):
        pass

    @abstractmethod
    def format_data(self):
        pass

    def print_report(self):
        print("Printing report...")

File: behaviorial/template-method/typescript/abstract.ts
export abstract class ReportGenerator {
	// Template method
	public generateReport(): void {
		this.collectData();
		this.processData();
		this.formatData();
		this.printReport();
	}

	// Abstract methods to be implemented by subclasses
	protected abstract collectData(): void;
	protected abstract processData(): void;
	protected abstract formatData(): void;

	// Final method that cannot be overridden
	private printReport(): void {
		console.log('Printing report...');
	}
}
